### PR TITLE
feat: Added Sample Hook for Enforcing APIGateway Authorizers

### PR DESCRIPTION
**Pull request by CoshUS**
*Issue #, if available:*
N/A

*Description of changes:*
Adding a sample hook for enforcing authorizers on all APIGateway resources and paths.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: hooks/python-hooks/apigw-enforce-authorizer/src/awssamples_apigwenforceauthorizer_hook/handlers.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    if validate_auth(target_name, resource_properties):
+        progress.status = OperationStatus.SUCCESS
+    else:
+        progress.status = OperationStatus.FAILED
+        progress.message = "Not all paths and methods have authorizers specified."
+    return progress
+
+
+@hook.handler(HookInvocationPoint.UPDATE_PRE_PROVISION)
```
Comment (mrinaudo-aws) [inline hooks/python-hooks/apigw-enforce-authorizer/src/awssamples_apigwenforceauthorizer_hook/handlers.py:50]: It seems that `pre_create_handler()` and `pre_update_handler()`, at this time, have the same code and behavior.  Can you please consider the following:
- move the decorator: `@hook.handler(HookInvocationPoint.UPDATE_PRE_PROVISION)` above, before the `@hook.handler(HookInvocationPoint.CREATE_PRE_PROVISION)` decorator in `pre_create_handler()`, and
- remove `pre_update_handler()` below.
Thank you!

File: hooks/python-hooks/apigw-enforce-authorizer/README.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# AWSSamples::APIGWEnforceAuthorizer::Hook
+
+This AWS CloudFormation Hook enforces all APIGateway resources to have authorizers attached when created and updated. 
+
+* When creating or updating an AWS::ApiGateway::RestApi or AWS::ApiGatewayV2::Api resource, the hook will check the JSON definition specified in `Body` to see whether all sections in `paths` have `security` defined.
+* When creating or updating an AWS::ApiGateway::Method or AWS::ApiGatewayV2::Route, the hook will check whether `AuthorizerId` is defined in the resource.
+
+## Limitations
+This hook serves as a simple example on how authorizers can be enforced on APIGateway resources. There are a few limitation associated with it in favor of simplicity. It is encouraged to fork this example and customize it depending on the need.
+1. OpenAPI definitions specified in S3 will be skipped and not validated.
```
Comment (mrinaudo-aws) [inline hooks/python-hooks/apigw-enforce-authorizer/README.md:10]:
```suggestion
1. OpenAPI definitions specified in an [Amazon Simple Storage Service](https://aws.amazon.com/s3/) (Amazon S3) bucket object will be skipped and not validated.
```

File: hooks/python-hooks/apigw-enforce-authorizer/awssamples-apigwenforceauthorizer-hook.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
+    "typeName": "AWSSamples::APIGWEnforceAuthorizer::Hook",
+    "description": "Enforces all APIGateway resources to have authorizers attached when created and updated.",
```
Comment (mrinaudo-aws) [inline hooks/python-hooks/apigw-enforce-authorizer/awssamples-apigwenforceauthorizer-hook.json:3]:
```suggestion
    "description": "Enforces Amazon API Gateway resources to have authorizers attached when created and updated.",
```

**Review by mrinaudo-aws — CHANGES_REQUESTED** [review]
Thank you!  Can you please go over suggestions and feedback, and then request another review when done?  Thanks!

File: hooks/python-hooks/apigw-enforce-authorizer/src/awssamples_apigwenforceauthorizer_hook/handlers.py (hunk 1 above)
Comment (CoshUS) [reply hooks/python-hooks/apigw-enforce-authorizer/src/awssamples_apigwenforceauthorizer_hook/handlers.py:30]: Moved both decorators to the same handler and removed the duplicated one.

**Comment by brianterry** [comment]
The contract test are failing. Please make sure that your test pass.

**Comment by CoshUS** [comment]
@brianterry @mrinaudo-aws 
Fixed contract tests. Please take a look when you have time. Thanks!
```
cfn test
=============================================== test session starts ===============================================
platform linux -- Python 3.7.13, pytest-7.0.1, pluggy-1.0.0 -- /usr/bin/python3.7
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/media/dev/Amazon/Workspace/aws-cloudformation-samples/hooks/python-hooks/apigw-enforce-authorizer/.hypothesis/examples')
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
rootdir: /tmp, configfile: pytest_61hxf2em.ini
plugins: hypothesis-6.37.2, random-order-1.0.4, localserver-0.5.1.post0
collected 21 items / 16 deselected / 5 selected                                                                   

../../../../../../../../tmp/hook/handler_pre_create.py::contract_pre_create_success PASSED                  [ 20%]
../../../../../../../../tmp/hook/handler_pre_create.py::contract_pre_create_failed PASSED                   [ 40%]
../../../../../../../../tmp/hook/handler_pre_update.py::contract_pre_update_success PASSED                  [ 60%]
../../../../../../../../tmp/hook/handler_pre_update.py::contract_pre_update_failed PASSED                   [ 80%]
../../../../../../../../tmp/resource/handler_misc.py::contract_check_asserts_work PASSED                    [100%]

================================================ warnings summary =================================================
../../../../../../../../home/cosh_/.local/lib/python3.7/site-packages/_pytest/config/__init__.py:1126
  /home/cosh_/.local/lib/python3.7/site-packages/_pytest/config/__init__.py:1126: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: hypothesis
    self._mark_plugins_for_rewrite(hook)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================== 5 passed, 16 deselected, 1 warning in 22.72s ===================================
=============================================== test session starts ===============================================
platform linux -- Python 3.7.13, pytest-7.0.1, pluggy-1.0.0 -- /usr/bin/python3.7
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/media/dev/Amazon/Workspace/aws-cloudformation-samples/hooks/python-hooks/apigw-enforce-authorizer/.hypothesis/examples')
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
rootdir: /tmp, configfile: pytest_motbmcte.ini
plugins: hypothesis-6.37.2, random-order-1.0.4, localserver-0.5.1.post0
collected 21 items / 16 deselected / 5 selected                                                                   

../../../../../../../../tmp/hook/handler_pre_create.py::contract_pre_create_success PASSED                  [ 20%]
../../../../../../../../tmp/hook/handler_pre_create.py::contract_pre_create_failed PASSED                   [ 40%]
../../../../../../../../tmp/hook/handler_pre_update.py::contract_pre_update_success PASSED                  [ 60%]
../../../../../../../../tmp/hook/handler_pre_update.py::contract_pre_update_failed PASSED                   [ 80%]
../../../../../../../../tmp/resource/handler_misc.py::contract_check_asserts_work PASSED                    [100%]

================================================ warnings summary =================================================
../../../../../../../../home/cosh_/.local/lib/python3.7/site-packages/_pytest/config/__init__.py:1126
  /home/cosh_/.local/lib/python3.7/site-packages/_pytest/config/__init__.py:1126: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: hypothesis
    self._mark_plugins_for_rewrite(hook)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================== 5 passed, 16 deselected, 1 warning in 21.42s ===================================
=============================================== test session starts ===============================================
platform linux -- Python 3.7.13, pytest-7.0.1, pluggy-1.0.0 -- /usr/bin/python3.7
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/media/dev/Amazon/Workspace/aws-cloudformation-samples/hooks/python-hooks/apigw-enforce-authorizer/.hypothesis/examples')
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
rootdir: /tmp, configfile: pytest_569am15q.ini
plugins: hypothesis-6.37.2, random-order-1.0.4, localserver-0.5.1.post0
collected 21 items / 16 deselected / 5 selected                                                                   

../../../../../../../../tmp/hook/handler_pre_create.py::contract_pre_create_success PASSED                  [ 20%]
../../../../../../../../tmp/hook/handler_pre_create.py::contract_pre_create_failed PASSED                   [ 40%]
../../../../../../../../tmp/hook/handler_pre_update.py::contract_pre_update_success PASSED                  [ 60%]
../../../../../../../../tmp/hook/handler_pre_update.py::contract_pre_update_failed PASSED                   [ 80%]
../../../../../../../../tmp/resource/handler_misc.py::contract_check_asserts_work PASSED                    [100%]

================================================ warnings summary =================================================
../../../../../../../../home/cosh_/.local/lib/python3.7/site-packages/_pytest/config/__init__.py:1126
  /home/cosh_/.local/lib/python3.7/site-packages/_pytest/config/__init__.py:1126: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: hypothesis
    self._mark_plugins_for_rewrite(hook)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================== 5 passed, 16 deselected, 1 warning in 21.57s ===================================
```

**Review by mrinaudo-aws — CHANGES_REQUESTED** [review]
Thank you for the update!  I cloned your fork, and ran contract tests from your feature branch.  At this time, I have 2 feedback items, as shown next.

### Item 1
Whilst 3 contract tests passed, 2 failed for me:
```
FAILED hook/handler_pre_create.py::contract_pre_create_success - AssertionError: status should be SUCCESS
FAILED hook/handler_pre_update.py::contract_pre_update_success - AssertionError: status should be SUCCESS
```
Can you please check as well, and follow up/fix this?

### Item 2
I ran the following command:
```
cfn generate && cfn submit --dry-run
```
and I realized there are differences with the `hook-role.yaml` file that is committed to the repo (please run `git diff` for more information after you run the command above).  Can you make sure you commit this updated file to the repo as well?

Thanks in advance!

**Comment by mrinaudo-aws** [comment]
> 

Please disregard Item 1 above - by mistake I ran `sam local start-lambda` from a different location.  My apologies.  I see contract tests are now actually working.

Thank you for updating the role-related template!

**Review by mrinaudo-aws — APPROVED** [review]
lgtm

**Pull request merged**